### PR TITLE
ATLAS-3655: Create 'spark_application' type to avoid 'spark_process' from being updated for multiple operations

### PR DESCRIPTION
**Pull request by vladhlinsky**
## What changes were proposed in this pull request?

Create `spark_application` type to avoid `spark_process` from being updated for multiple operations. Currently, Spark Atlas Connector uses `spark_process` as a top-level type for a Spark session, thus it's being updated for multiple operations within the same session.

The following statements:
```
spark.sql("create table table_1(col1 int,col2 string)");
spark.sql("create table table_2 as select * from table_1");
```
result in the next correct lineage:
```
table1 ------> spark_process1 -------> table2
```
but executing similar statements in the same spark session:
```
spark.sql("create table table_3(col1 int,col2 string)"); 
spark.sql("create table table_4 as select * from table_3");
```
result in the same `spark_process` being updated and the lineage now connects all the 4 tables.
The proposal is to create a `spark_application` entity and associate all `spark_process` entities (created within that session) to it.

## How was this patch tested?

Manually using modified version of Spark Atlas Connector:
- Installed and started Atlas.
- Executed the next statements using spark-shell:

```
spark.sql("create table table_1_17(col1 int,col2 string)");
spark.sql("create table table_2_17 as select * from table_1_17");
spark.sql("create table table_3_17(col1 int,col2 string)");
spark.sql("create table table_4_17 as select * from table_3_17");
```

- Verified that all 4 entites are connected in Atlas lineage.
- `1100-spark_model.json` is updated with proposed changes.
- Once again executed similar statements:

```
spark.sql("create table table_1_37(col1 int,col2 string)");
spark.sql("create table table_2_37 as select * from table_1_37");
spark.sql("create table table_3_37(col1 int,col2 string)");
spark.sql("create table table_4_37 as select * from table_3_37");
```

- Verified that two `spark_process` entities are created,
that have a single `spark_application` entity as `application`.
Each of these processes has it's own lineage.

### CODE REVIEW

**Comment by vladhlinsky** [comment]
Attaching screenshots.
- Installed and started Atlas.
- Executed the next statements using spark-shell:

```
spark.sql("create table table_1_17(col1 int,col2 string)");
spark.sql("create table table_2_17 as select * from table_1_17");
spark.sql("create table table_3_17(col1 int,col2 string)");
spark.sql("create table table_4_17 as select * from table_3_17");
```
- Verified that all 4 entites are connected in Atlas lineage.
Screenshot from 2020-02-27 19-31-09

- `1100-spark_model.json` is updated with proposed changes.
- Once again executed similar statements:

```
spark.sql("create table table_1_37(col1 int,col2 string)");
spark.sql("create table table_2_37 as select * from table_1_37");
spark.sql("create table table_3_37(col1 int,col2 string)");
spark.sql("create table table_4_37 as select * from table_3_37");
```

- Verified that two `spark_process` entities are created,
that have a single `spark_application` entity as `application`.
Each of these processes has it's own lineage.

Screenshot from 2020-03-04 23-16-44
Screenshot from 2020-03-04 23-17-02
Screenshot from 2020-03-04 23-17-10
Screenshot from 2020-03-04 23-17-34
Screenshot from 2020-03-04 23-19-56

**Comment by vladhlinsky** [comment]
cc @HeartSaVioR @sarathsubramanian

File: addons/models/1000-Hadoop/1100-spark_model.json [hunk 1]
```diff
@@ -305,6 +305,34 @@
...
+    {
+      "name": "spark_application",
+      "superTypes": [
+        "Process"
+      ],
+      "serviceType": "spark",
+      "typeVersion": "1.0",
+      "attributeDefs": [
+        {
+          "name": "currUser",
```
Comment (sarathsubramanian) [inline addons/models/1000-Hadoop/1100-spark_model.json]: consider renaming "currUser" to "currentUser"

File: addons/models/1000-Hadoop/1100-spark_model.json [hunk 2]
```diff
@@ -470,6 +498,24 @@
         "cardinality": "SINGLE"
       },
       "propagateTags": "NONE"
+    },
+    {
+      "name": "spark_application_process",
```
Comment (sarathsubramanian) [inline addons/models/1000-Hadoop/1100-spark_model.json]: "spark_application_process" => "spark_application_processes"

File: addons/models/1000-Hadoop/1100-spark_model.json [hunk 3]
```diff
@@ -305,6 +305,34 @@
...
+      "serviceType": "spark",
+      "typeVersion": "1.0",
+      "attributeDefs": [
+        {
+          "name": "currUser",
+          "typeName": "string",
+          "isOptional": true,
+          "cardinality": "SINGLE",
+          "isUnique": false,
+          "isIndexable": false,
```
Comment (sarathsubramanian) [inline addons/models/1000-Hadoop/1100-spark_model.json]: isIndexable => true

File: addons/models/1000-Hadoop/1100-spark_model.json [hunk 4]
```diff
@@ -305,6 +305,34 @@
...
+          "searchWeight": 10
+        },
+        {
+          "name": "remoteUser",
+          "typeName": "string",
+          "isOptional": true,
+          "cardinality": "SINGLE",
+          "isUnique": false,
+          "isIndexable": false,
```
Comment (sarathsubramanian) [inline addons/models/1000-Hadoop/1100-spark_model.json]: isIndexable => true

File: addons/models/1000-Hadoop/1100-spark_model.json (hunk 1 above)
Comment (vladglinsky) [reply addons/models/1000-Hadoop/1100-spark_model.json]: Thanks! Changed to  "currentUser".

Comment (vladhlinsky) [reply addons/models/1000-Hadoop/1100-spark_model.json]: Thanks! Changed to "currentUser".

File: addons/models/1000-Hadoop/1100-spark_model.json (hunk 2 above)
Comment (vladhlinsky) [reply addons/models/1000-Hadoop/1100-spark_model.json]: Renamed.

File: addons/models/1000-Hadoop/1100-spark_model.json (hunk 3 above)
Comment (vladhlinsky) [reply addons/models/1000-Hadoop/1100-spark_model.json]: Changed. Thanks!

File: addons/models/1000-Hadoop/1100-spark_model.json (hunk 4 above)
Comment (vladhlinsky) [reply addons/models/1000-Hadoop/1100-spark_model.json]: Changed.

**Comment by sarathsubramanian** [comment]
changes looks good. +1

**Comment by fpompermaier** [comment]
Aren't the 2 names of  "type": "spark_application",  "name": "application" and "type": "spark_process", "name": "application" swapped?

**Comment by vladhlinsky** [comment]
@fpompermaier thank you for your comment.

No, they are not swapped. A relationship definition specifies relationship attribute on a type. This means that `spark_application_processes` define `application` relationship attribute on the `spark_process` type and `processes` attribute on the `spark_application` type. [Please, refer screenshots for an example.](https://github.com/apache/atlas/pull/91#issuecomment-595413075)

**Comment by fpompermaier** [comment]
> @fpompermaier thank you for your comment.
> 
> No, they are not swapped. A relationship definition specifies relationship attribute on a type. This means that `spark_application_processes` define `application` relationship attribute on the `spark_process` type and `processes` attribute on the `spark_application` type. [Please, refer screenshots for an example.](https://github.com/apache/atlas/pull/91#issuecomment-595413075)

Ok! Thanks fort the clarification! I'm quite new to Atlas

**Comment by nixonrodrigues** [comment]
@vladhlinsky ,

 this PR has conflicts, can you please rebase with master and update PR.

**Pull request merged**